src/framework/theme/components/autocomplete/autocomplete.directive.ts

```diff
@@ -0,0 +1,417 @@
+
+  protected overlayRef: NbOverlayRef;
+
+  protected keyManager: NbActiveDescendantKeyManager<NbOptionComponent<T>>;
+
+  protected destroy$: Subject<boolean> = new Subject<boolean>();
+
+  /**
+   * Determines is autocomplete opened.
+   * */
+  get isOpen(): boolean {
+    return this.overlayRef && this.overlayRef.hasAttached();
+  }
+
+  /**
+   * Determines is autocomplete overlay closed.
+   * */
+  get isClosed(): boolean {
+    return !this.isOpen;
+  }
+
+  /**
+   * Provides autocomplete component.
+   * */
+  @Input('nbAutocomplete')
+  set setAutocomplete(autocomplete: NbAutocompleteComponent<T>) {
+    this.autocomplete = autocomplete;
+  }
+
+  @HostBinding('class.nb-autocomplete-position-top')
+  get top(): boolean {
+    return this.isOpen && this.autocomplete.options.length && this.autocomplete.overlayPosition === NbPosition.TOP;
+  }
+
+  @HostBinding('class.nb-autocomplete-position-bottom')
+  get bottom(): boolean {
+    return this.isOpen && this.autocomplete.options.length && this.autocomplete.overlayPosition === NbPosition.BOTTOM;
+  }
+
+  @HostBinding('attr.role')
+  role: string = 'combobox';
+
+  @HostBinding('attr.aria-autocomplete')
+  ariaAutocomplete: string = 'list';
+
+  @HostBinding('attr.haspopup')
+  hasPopup: string = 'true';
+
+  @HostBinding('attr.aria-expanded')
+  get ariaExpanded(): string {
+    return this.isOpen && this.isOpen.toString();
+  }
+
+  @HostBinding('attr.aria-owns')
+  get ariaOwns() {
+    return this.isOpen ? this.autocomplete.id : null;
+  }
+
+  @HostBinding('attr.aria-activedescendant')
+  get ariaActiveDescendant() {
+    return (this.isOpen && this.keyManager.activeItem) ? this.keyManager.activeItem.id : null;
+  }
+
+  constructor(
+    protected hostRef: ElementRef,
+    protected overlay: NbOverlayService,
+    protected cd: ChangeDetectorRef,
+    protected triggerStrategyBuilder: NbTriggerStrategyBuilderService,
+    protected positionBuilder: NbPositionBuilderService,
+    protected activeDescendantKeyManagerFactoryService:
+      NbActiveDescendantKeyManagerFactoryService<NbOptionComponent<T>>) {}
+
+  ngAfterViewInit() {
+    this.triggerStrategy = this.createTriggerStrategy();
+    this.subscribeOnTriggers();
+  }
+
+  ngOnDestroy() {
+
+    if (this.triggerStrategy) {
+      this.triggerStrategy.destroy();
+    }
+
+    if (this.positionStrategy) {
+      this.positionStrategy.dispose();
+    }
+
+    if (this.overlayRef) {
+      this.overlayRef.dispose();
+    }
+
+    this.destroy$.next(true);
+    this.destroy$.complete();
+  }
+
+  @HostListener('input')
+  protected handleInput() {
+    const currentValue = this.hostRef.nativeElement.value;
+    this._onChange(currentValue);
+    this.setHostInputValue(this.getDisplayValue(currentValue));
+
+    // Detach overlay after content updated if there is no options.
+    setTimeout(() => {
+      this.autocomplete.options.length ? this.show() : this.hide();
+    });
+  }
+
+  @HostListener('keydown.arrowDown', ['$event'])
```

Comment:
We don't expect event parameter in `handleKeydown`. Can't `['$event']` be removed?

Reply:
Yes, removed.